%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{ELMER: Inferring Regulatory Element Landscapes and Transcription Factor Networks Using Methylomes}
% !Rnw weave = knitr
\documentclass{article}
<<style-knitr, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@
\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{grffile}
\usepackage{float}

\title{ELMER: An R/Bioconductor Tool \\
Inferring Regulatory Element Landscapes and Transcription Factor Networks Using Methylomes}
\author {Lijing Yao [cre, aut], 
Ben Berman [aut], 
Peggy Farnham [aut]\\ 
Hui Shen [ctb], 
Peter Laird [ctb], 
Simon Coetzee [ctb]}

\begin{document}
<<setup, include=FALSE, cache=FALSE>>=
library(knitr)
opts_chunk$set(fig.align='center', cache=TRUE,par=TRUE)
knit_hooks$set(par=function(before, options, envir){
if (before && options$fig.show!='none') par(mar=c(1,1,.1,.1),cex.lab=1,cex.axis=1,mgp=c(2,.7,0),tcl=-.3)
}, crop=hook_pdfcrop)
@
\maketitle

\tableofcontents
\section{Introduction}
This document provides an introduction of the \Rpackage{ELMER}, which is designed 
to use DNA methylation and gene expression data sets from a large number
of tissue samples to infer regulatory element landscapes and transcription factor network.
It includes functions for identifying probes at distal regulatory regions with differential 
DNA methylation levels, predicting genes whose expression associates with the differentially
methylated probes and discovering the functional regulatory TFs. This package can be easily
applied to TCGA public available cancer data sets and to custom DNA methylation 
and gene expression data sets. 

\subsection{Installing and loading ELMER}
To obtain a copy of \Rpackage{ELMER}, you will need to install \Rpackage{devtools}
<<installing, eval=FALSE>>==
install.packages(devtools)
library(devtools);
devtools::install_github("lijingya/ELMER");
@

\section{Download example data}
The following steps can be used to download the example data set for \Rpackage{ELMER}
<<example.data.run, echo=FALSE, hide=TRUE, message=FALSE, include=FALSE>>=
if(!file_test("-d", "./ELMER.example")) {
  URL <- "https://dl.dropboxusercontent.com/u/61961845/ELMER.example.tar.gz"
  download.file(URL,destfile = "ELMER.example",method= "curl")
  untar("./ELMER.example")
}
library(ELMER, quietly = TRUE)
@

<<example.data, eval=FALSE>>=
#Example file download from URL: https://dl.dropboxusercontent.com/u/61961845/ELMER.example.tar.gz
URL <- "https://raw.githubusercontent.com/lijingya/ELMER/master/ELMER.example.tar.xz"
download.file(URL,destfile = "ELMER.example",method= "curl")
untar("./ELMER.example")
library(ELMER)
@


\section{Quick start: running TCGA example}
A function, TCGA.pipe, is the easy usage for downloading TCGA data and performing all 
the analyses in ELMER. For illustration purpose, we skip the downloading step. 
The user can use the getTCGA function to download TCGA data or
use TCGA.pipe by including "download" in the analysis option.

<<tcga.pipe, cache=TRUE>>=
TCGA.pipe("LUSC",wd="./ELMER.example",cores=detectCores()/2,permu.size=300,
          analysis = c("distal.enhancer","diffMeth","pair","motif","TF.search"),
          diff.dir="hypo",rm.chr=paste0("chr",c(1:22,"X","Y")))
@


\section{Input data}

The whole pipeline analyses in \Rpackage{ELMER} needs at least 4 input files:
a matrix of DNA methylation from HM450K platform; a matrix of gene expression for the 
same samples; a GRanges object containing the information for probes on HM450K such as names 
and coordinates; a gene annotation which is also a GRanges object. When TCGA data are used,
the sample information will be automatically generated by fetch.mee function. However
sample information should be provided when using custom data.

\subsection{DNA methylation data}
Raw DNA methylation data can be processed by \Rpackage{Methylumi} or \Rpackage{minfi}
generating DNA methylation information for each CpG. The DNA methylation level 
at each CpG is referred to as a beta ($\beta$) value, calculated as (M/(M+U)), 
where M represents the methylated allele intensity and U the unmethylated allele 
intensity. Beta values range from 0 to 1, reflecting the fraction of methylated 
alleles at each CpG in the each tumor; beta values close to 0 indicates low 
levels of DNA methylation and beta values close to 1 indicates high levels 
of DNA methylation. Generate a matrix with DNA methylation beta values for all
the samples (columns) and probe loci (rows) and save matrix as meth.rda

<<dna.methylation.data, cache=TRUE>>=
load("./ELMER.example/Result/LUSC/LUSC_meth_refined.rda")
Meth[1:10, 1:2]
@

\subsection{Gene expression data}
Gene expression values can be generated from different platforms such as array or 
RNA-seq, gene level or transcript level gene expression calling. Generate a matrix 
with gene expression values for all the samples (columns) and genes (rows) 
and save matrix as RNA.rda

<<gene.expression.data, cache=TRUE>>=
load("./ELMER.example/Result/LUSC/LUSC_RNA_refined.rda")
GeneExp[1:10, 1:2]
@


\subsection{Sample information}
Sample information should be stored as a data.frame object containing sample ID, 
group labels (such as tumor, normal) and other description for each sample. 
When TCGA data were used, tumor, normal group label will be automatically generated by 
fetch.mee function by specifying option TCGA=TRUE.

<<sample.information, cache=TRUE>>=
mee <- fetch.mee(meth=Meth, exp=GeneExp, TCGA=TRUE)
head(getSample(mee))
@

\subsection{Probe information}
Probe information should be stored as a GRanges object containing the coordinate 
of each probe on the DNA methylation array and names of each probe. 
The default probe information is for HM450K. 

<<probe.information, cache=TRUE>>=
probe <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19, what="Locations")
probe <- GRanges(seqnames=probe$chr,
		ranges=IRanges(probe$pos,
			width=1,
			names=rownames(probe)),
		strand=probe$strand,
		name=rownames(probe))
mee <- fetch.mee(meth=Meth, exp=GeneExp, TCGA=TRUE, probeInfo=probe)
getProbeInfo(mee)
@

\subsection{Gene information}
Gene information should be stored as a GRanges object containing coordinates of 
each gene, gene id, gene symbol and gene isoform id. The default gene information 
is the UCSC gene annotation. 


<<gene.information, cache=TRUE>>=
load(system.file("extdata","UCSC_gene_hg19.rda",package = "ELMER"))
## In TCGA expression data, geneIDs were used as the rowname for each row. However, numbers 
## can't be the rownames, "ID" was added to each gene id functioning as the rowname.
## If your geneID is consistent with the rownames of the gene expression matrix, adding "ID" 
## to each geneID can be skipped.
txs$GENEID <- paste0("ID",txs$GENEID)            
geneInfo <- promoters(txs,upstream = 0, downstream = 0)
save(geneInfo,file="./ELMER.example/Result/LUSC/geneAnnot.rda")
mee <- fetch.mee(meth=Meth, exp=GeneExp, TCGA=TRUE, geneInfo=txs)
getGeneInfo(mee)
@

\subsection{MEE.data}
The above 5 components will generate a MEE.data object as the main input for mulitple functions 
in \Rpackage{ELMER}.

<<mee.data, cache=TRUE>>=
mee <- fetch.mee(meth=Meth, exp=GeneExp, TCGA=TRUE, probeInfo=probe, geneInfo=txs)
mee
@

\section{Illustration of ELMER analysis steps}
A subset of chromosome 1 data from TCGA LUSC were used as illustruation. 

\subsection{Selection of probes within biofeatures}
A function, get.feature.probe, is used to select probes that are located within 
biofeatures such as H3K27ac ChIP-seq peaks. As default, the get.feature.probe function
will automatically select distal enhancer probes on HM450K which are at least 
2kb away from the TSS annotated by GENCODE V15 and UCSC-gene and locate within 
the putative comprehensive enhancers from REMC, ENCODE and FANTOM5. 
<<selection.of.probes.within.biofeatures, cache=TRUE>>=
#get distal enhancer probes that are 2kb away from TSS and overlap with REMC and FANTOM5 
#enhancers on chromosome 1
Probe <- get.feature.probe(probe=probe, rm.chr=paste0("chr",c(2:22,"X","Y")))
save(Probe,file="./ELMER.example/Result/LUSC/probeInfo_feature.rda")
@


\subsection{Identifying differentially methylated probes}
A function, get.diff.meth, will be used to identify differentially methylated 
probes among the ones within biofeatures, which are selected in the above step. 
<<identifying.differentially.methylated.probes, cache=TRUE>>=
## fetch.mee can take path as input.
mee <- fetch.mee(meth="./ELMER.example/Result/LUSC/LUSC_meth_refined.rda",
                 exp="./ELMER.example/Result/LUSC/LUSC_RNA_refined.rda", TCGA=TRUE, 
                 probeInfo="./ELMER.example/Result/LUSC/probeInfo_feature.rda", 
                 geneInfo="./ELMER.example/Result/LUSC/geneAnnot.rda") 

sig.diff <- get.diff.meth(mee, cores=detectCores()/2, dir.out ="./ELMER.example/Result/LUSC", 
                          diff.dir="hypo", pvalue = 0.01)

sig.diff$hypo[1:10,]   ## significantly hypomethylated probes

# get.diff.meth automatically save output files. 
# getMethdiff.hypo.probes.csv contains statistics for all the probes.
# getMethdiff.hypo.probes.significant.csv contains only the significant probes.
dir(path = "./ELMER.example/Result/LUSC", pattern = "getMethdiff")  
@


\subsection{Identifying putative probe-gene pairs}
A function, get.pair function, will be used to identify putative target genes for selected probes. 
This step is the most time consuming step since it requires a large amount calculations 
for permutation. The greater the permutation time is, the longer it will take. 
It is recommended to use multiple cores for this step. Default permutation time 
is 1000 which may need 12 hrs by 4 cores. However 10,000 permutations is recommended 
which may cost 2 days, if high confidence results are desired.
<<identifying.putative.probe.gene.pairs, cache=TRUE>>=
### identify target gene for significantly hypomethylated probes.

Sig.probes <- read.csv("./ELMER.example/Result/LUSC/getMethdiff.hypo.probes.significant.csv",
                       stringsAsFactors=FALSE)[,1]  
head(Sig.probes)  # significantly hypomethylated probes

## Collect nearby 20 gene for Sig.probes
nearGenes <-GetNearGenes(TRange=getProbeInfo(mee,probe=Sig.probes),
                         geneAnnot=getGeneInfo(mee),cores=detectCores()/2)

## Identify significant probe-gene pairs
Hypo.pair <-get.pair(mee=mee,probes=Sig.probes,nearGenes=nearGenes,
                     permu.dir="./ELMER.example/Result/LUSC/permu",permu.size=300,Pe = 0.01,
                     dir.out="./ELMER.example/Result/LUSC",cores=detectCores()/2,label= "hypo")

head(Hypo.pair)  ## significant probe-gene pairs

# get.pair automatically save output files. 
#getPair.hypo.all.pairs.statistic.csv contains statistics for all the probe-gene pairs.
#getPair.hypo.pairs.significant.csv contains only the significant probes.
dir(path = "./ELMER.example/Result/LUSC", pattern = "getPair") 
@


\subsection{Motif enrichment analysis on the selected probes}
A function, get.enriched.motif, will be used to calculate enrichment of the motifs from 
factorbook and JASPER for the selected probes. Odds Ratio is used to assess 
the enrichment levels and 95\% confidence interval of Odds Ratio is calculated.
<<motif.enrichment.analysis.on.selected.probes, cache=TRUE>>=
### identify enriched motif for significantly hypomethylated probes which 
##have putative target genes.

Sig.probes.paired <- read.csv("./ELMER.example/Result/LUSC/getPair.hypo.pairs.significant.csv",
                              stringsAsFactors=FALSE)[,1]  
head(Sig.probes.paired) # significantly hypomethylated probes with putative target genes

enriched.motif <-get.enriched.motif(probes=Sig.probes.paired, 
                                    dir.out="./ELMER.example/Result/LUSC", label="hypo",
                                    min.incidence = 10,lower.OR = 1.1)
names(enriched.motif)  # enriched motifs

# get.enriched.motif automatically save output files. 
# getMotif.hypo.enriched.motifs.rda contains enriched motifs and the probes with the motif. 
# getMotif.hypo.motif.enrichment.csv contains summary of enriched motifs.
dir(path = "./ELMER.example/Result/LUSC", pattern = "getMotif") 

# motif enrichment figure will be automatically generated.
dir(path = "./ELMER.example/Result/LUSC", pattern = "motif.enrichment.pdf") 
@


\subsection{Identifying regulatory TF}
A function, get.TFs, will use the anti-correlation of a particular TF and the level of 
demethylation at its binding sites to predict the regulatory TF.

<<identifying.regulatory.tf, cache=TRUE>>=
### identify regulatory TF for the enriched motifs

load("./ELMER.example/Result/LUSC/getMotif.hypo.enriched.motifs.rda")
TF <- get.TFs(mee=mee, enriched.motif=enriched.motif,dir.out="./ELMER.example/Result/LUSC", 
              cores=detectCores()/2, label= "hypo")

# get.TFs automatically save output files. 
# getTF.hypo.TFs.with.motif.pvalue.rda contains statistics for all TF with average 
# DNA methylation at sites with the enriched motif.
# getTF.hypo.significant.TFs.with.motif.summary.csv contains only the significant probes.
dir(path = "./ELMER.example/Result/LUSC", pattern = "getTF")  

# TF ranking plot based on statistics will be automatically generated.
dir(path = "./ELMER.example/Result/LUSC/TFrankPlot", pattern = "pdf") 
@
\begin{minipage}{\textwidth}
\section{Generating figures}
\subsection{Scatter plots}
Generate scatter plots for one probes' nearby 20 gene expression vs DNA methylation at this probe. Figure \ref{fig:figure1}

\begin{figure}[H]
<<figure1,fig.width=11, fig.height=10,fig.scap="Scatter Plot of 20 nearby genes",cache=TRUE>>=
scatter.plot(mee,byProbe=list(probe=c("cg19403323"),geneNum=20), 
             category="TN", dir.out ="./ELMER.example/Result/LUSC", save=FALSE) 
@
\protect\caption{SEach scatter plot shows the methylation level of an example 
probe cg19403323 in all LUSC samples plotted against the expression of one of 
20 adjacent genes.
\label{fig:figure1}}
\end{figure}
\end{minipage}

\begin{minipage}{\textwidth}
\subsubsection{Scatter Plot of One Pair}
Generate a scatter plot for one probe-gene pair. Figure \ref{fig:figure2}

\begin{figure}[H]
<<figure2, fig.height=4,fig.scap="Scatter Plot of One Pair", cache=TRUE>>=
scatter.plot(mee,byPair=list(probe=c("cg19403323"),gene=c("ID255928")), 
             category="TN", save=FALSE,lm_line=TRUE) 
@
\protect\caption{Scatter plot shows the methylation level of an example probe 
cg19403323 in all LUSC samples plotted against the expression of the putative 
target gene SYT14.
\label{fig:figure2}}
\end{figure}
\end{minipage}

\begin{minipage}{\textwidth}
\subsubsection{TF expression vs. average DNA methylation}
Generate scatter plot for TF expression vs average DNA methylation of the sites 
with certain motif. Figure \ref{fig:figure3}
\begin{figure}[H]
<<figure3, fig.height=4, fig.scap="TF expression vs. average DNA methylation", cache=TRUE>>=
load("ELMER.example/Result/LUSC/getMotif.hypo.enriched.motifs.rda")
scatter.plot(mee,byTF=list(TF=c("TP53","TP63","TP73"),
             probe=enriched.motif[["TP53"]]), category="TN",
             save=FALSE,lm_line=TRUE)
@
\protect\caption{Each scatter plot shows the average 
methylation level of sites with the TP53 motif in all LUSC samples plotted 
against the expression of the transcription factor TP53, TP63, TP73 respectively.
\label{fig:figure3}}
\end{figure}
\end{minipage}

\subsection{Schematic plot}
Schematic plot shows a breif view of linkages between genes and probes. To make a 
schematic plot, "Pair" object should be generated first.
<<makepair, cache=TRUE>>=
# Make a "Pair" object for schematic.plot
pair <- fetch.pair(pair="./ELMER.example/Result/LUSC/getPair.hypo.pairs.significant.withmotif.csv",
                   probeInfo = "./ELMER.example/Result/LUSC/probeInfo_feature.rda",
                   geneInfo = "./ELMER.example/Result/LUSC/geneAnnot.rda")
@
\begin{minipage}{\textwidth}
\subsubsection{Nearby Genes}
Generate schematic plot for one probe with 20 nearby genes and label 
the gene significantly linked with the probe in red. Figure \ref{fig:figure4}

\begin{figure}[H]
<<figure4,fig.width=8, fig.height=3,fig.scap="Nearby Genes", cache=TRUE>>=
schematic.plot(pair=pair, byProbe="cg19403323",save=FALSE)
@
\protect\caption{The schematic plot shows probe colored 
in blue and the location of nearby 20 genes. The genes significantly linked to the probe 
were shown in red.
\label{fig:figure4}}
\end{figure}

\subsubsection{Nearby Probes}
Generate schematic plot for one gene with the probes which the gene is significantly 
linked to. Figure \ref{fig:figure5}
\begin{figure}[H]
<<figure5,fig.width=8, fig.height=3,fig.scap="Nearby Probes", cache=TRUE>>=
schematic.plot(pair=pair, byGene="ID255928",save=FALSE)
@
\protect\caption{The schematic plot shows the gene 
colored in red and all blue colored probes, which are significantly linked to the 
expression of this gene.
\label{fig:figure5}}
\end{figure}
\end{minipage}

\begin{minipage}{\textwidth}
\subsection{Motif enrichment plot}
Motif enrichment plot shows the enrichment levels for the selected motifs. Figure\ref{fig:figure6}
\begin{figure}[H]
<<figure6, fig.height=2,fig.scap="Motif Enrichment", cache=TRUE>>=
motif.enrichment.plot(motif.enrichment="./ELMER.example/Result/LUSC/getMotif.hypo.motif.enrichment.csv", 
                      significant=list(OR=1.3,lowerOR=1.3), dir.out ="ELMER.example/Result/LUSC",
                      label="hypo", save=FALSE)  ## different signficant cut off.
@
\protect\caption{The plot shows the Odds Ratio (x axis) 
for the selected motifs with OR above 1.3 and lower boundary of OR above 1.3. 
The range shows the 95\% confidence interval for each Odds Ratio.
\label{fig:figure6}}
\end{figure}


\subsection{TF ranking plot}
TF ranking plot shows statistic -log10(P value) assessing the anti-correlation level 
of TFs expression level with average DNA methylation level at sites with a given motif. Figure \ref{fig:figure7}

\begin{figure}[H]
<<figure7,fig.width=5.5, fig.height=3.5,fig.scap="TF Ranking", cache=TRUE>>=
load("./ELMER.example/Result/LUSC/getTF.hypo.TFs.with.motif.pvalue.rda")
TF.rank.plot(motif.pvalue=TF.meth.cor, motif="TP53", TF.label=list(TP53=c("TP53","TP63","TP73")),
            save=FALSE) 
@

\protect\caption{Shown are TF ranking plots based on 
the score (-log(P value)) of association between TF expression and DNA methylation of 
the TP53 motif in the LUSC cancer type . The dashed line indicates the boundary 
of the top 5\% association score. The top 3 associated TFs and the TF family members
(dots in red) that are associated with that specific motif are labeled in the plot.
\label{fig:figure7}}
\end{figure}
\end{minipage}
\newpage
<<finalsession, cache=TRUE>>=
sessionInfo()
@

\end{document}
